proxy/src/watch_service.rs

```diff
@@ -0,0 +1,51 @@
+use futures::{Async, Poll, Stream};
+use futures_watch::Watch;
+use tower_service::Service;
+
+pub trait Rebind<T> {
+    type Service: Service;
+    fn rebind(&mut self, t: &T) -> Self::Service;
+}
+
+/// A Service that updates itself as a Watch updates.
+#[derive(Debug)]
+pub struct WatchService<T, R: Rebind<T>> {
+    watch: Watch<T>,
+    rebind: R,
+    inner: R::Service,
+}
+
+impl<T, R: Rebind<T>> WatchService<T, R> {
+    pub fn new(watch: Watch<T>, mut rebind: R) -> WatchService<T, R> {
+        let inner = rebind.rebind(&*watch.borrow());
+        WatchService { watch, rebind, inner }
+    }
+}
+
+impl<T, R: Rebind<T>> Service for WatchService<T, R> {
+    type Request = <R::Service as Service>::Request;
+    type Response = <R::Service as Service>::Response;
+    type Error = <R::Service as Service>::Error;
+    type Future = <R::Service as Service>::Future;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        // Check to see if the watch has been updated and, if so, rebind the service.
+        if let Async::Ready(Some(())) = self.watch.poll().expect("watch poll") {
```

Comment:
I think we should handle the error case; i.e. use `try_ready!()`.

Reply:
I'm not convinced it's necessary or desirable.

To do this, we'd have to introduce a new `enum Error<E> { WatchError, Inner(E) }`, since we can generally turn a WatchError into a Self::Error.  Furthermore, `watch.poll()` can't actually fail... https://github.com/carllerche/better-future/blob/07baa13e91fefe7a51533dfde7b4e69e109ebe14/futures-watch/src/lib.rs#L288-L307

So, this comes with the cost of boilerplate, and no functional benefit.

Reply:
OK, I think it's good to document that assumption for if/when we change this code to depend on something that can fail.

Reply:
I changed this to remove the `expect`, and instead only match on `Ok(_)` (with a comment that the poll can't actually fail).